src/hotspot/share/services/nmt/memoryLogRecorder.cpp

```diff
@@ -0,0 +1,961 @@
+  fprintf(stderr, "native malloc used following distinct allocation sizes: ");
+  for (size_t i = 0; i < _malloc_requests_count; i++) {
+    if ((histogram_actual_sizes[i] > 0) && (histogram_actual_sizes[i] > actual_sizes_last)) {
+      actual_sizes_count++;
+      actual_sizes_last = histogram_actual_sizes[i];
+      fprintf(stderr, "%zu ", histogram_actual_sizes[i]);
+    }
+  }
+  fprintf(stderr, "\n\n");
+  fprintf(stderr, "native malloc used %zu distinct allocation sizes\n", actual_sizes_count);
+
+//  assert(r_total == total_requested, "r_total:%zu == total_requested:%zu", r_total, total_requested);
+//  assert(a_total == total_actual, "a_total:%zu == total_actual:%zu", a_total, total_actual);
+//  assert(a_total-r_total == o_total, "a_total-r_total:%zu == o_total:%zu", a_total-r_total, o_total);
+//  assert(alloc_overhead == o_total, "alloc_overhead:%zu == total_requested:%zu", alloc_overhead, o_total);
+
+  ALLOW_C_FUNCTION(::free, ::free(histogram_actual_sizes);)
+  ALLOW_C_FUNCTION(::free, ::free(histogram_counts);)
+}
+
+void NMT_MemoryLogRecorder::print_entry(Entry* e) {
+  if (e != nullptr) {
+    fprintf(stderr, "{ %18ld, %18p, %18p", e->time, e->ptr, e->old);
+    for (int i = 0; i < NMT_TrackingStackDepth; i++) {
+      fprintf(stderr, ", %18p", e->stack[i]);
+    }
+    fprintf(stderr, ", %7u, %7u, %7u, %1d", (unsigned)e->requested, (unsigned)e->actual, (unsigned)e->flags, is_active(e));
+    fprintf(stderr, ", \"%s\"},\n", NMTUtil::flag_to_name(e->flags));
+  } else {
+    fprintf(stderr, "nullptr\n");
+  }
+}
+
+void NMT_MemoryLogRecorder::print_records(Entry* entries, size_t count) {
+  for (size_t c = 0; c < count; c++) {
+    Entry* e = &entries[c];
+    print_entry(e);
+  }
+}
+
+void NMT_MemoryLogRecorder::report_by_component(Entry* entries, size_t count) {
+  fprintf(stderr, "\n");
+  fprintf(stderr, "--------------------------------------------------------------------------------------------\n");
+  fprintf(stderr, "         component name:  mallocs: reallocs:    frees:   requested:   allocated:   overhead:\n");
+  fprintf(stderr, "                          (count)   (count)    (count)    (bytes)      (bytes)   (mem diff %%)\n");
+  fprintf(stderr, "--------------------------------------------------------------------------------------------\n");
+
+  size_t total_requested = 0;
+  size_t total_actual = 0;
+  for (size_t c = 0; c < count; c++) {
+    Entry* e = access_active(entries, c);
+    if ((e != nullptr) && (is_alloc(e))) {
+      total_requested += e->requested;
+      total_actual += e->actual;
+    }
+  }
+  size_t alloc_overhead = (total_actual - total_requested);
+
+  for (int i = 0; i < mt_number_of_types; i++) {
+    size_t mallocs = 0;
+    size_t reallocs = 0;
+    size_t frees = 0;
+    size_t requested = 0;
+    size_t allocated = 0;
+    for (size_t c = 0; c < count; c++) {
+      Entry* e = &entries[c];
+      if (e->flags == NMTUtil::index_to_flag(i)) {
+        if (is_active(e)) {
```

Comment:
Is nothing done if it's not active?

Then, why not do:

```c++
if (!is_active(e)) continue;
```

Same for the flag check above.